File: src/main/java/org/modelingvalue/nelumbo/logic/Quantifier.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.logic;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.lang.Type;
import org.modelingvalue.nelumbo.lang.Variable;

public abstract class Quantifier extends CompoundPredicate {

    @Serial
    private static final long serialVersionUID = -4838100281214165385L;

    protected Quantifier(Functor functor, List<AstElement> elements, Node declaration, Object... args) {
        super(functor, elements, declaration, args);
    }

    protected Quantifier(Object[] args, Node functorOrType, List<AstElement> elements, Quantifier declaration) {
        super(args, functorOrType, elements, declaration);
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<Variable> localVars() {
        return (List<Variable>) get(0);
    }

    public final Predicate predicate() {
        return predicate(1);
    }

    @Override
    protected int countNrOfUnbound() {
        return (int) getBinding().removeAllKey(localVars()).filter(e -> e.getValue() instanceof Type).count();
    }

    @Override
    protected boolean doGetBinding(Object varVal, int i) {
        return i > 0;
    }

    @Override
    protected boolean doSetBinding(Object varVal, int i) {
        return i > 0 || varVal instanceof Variable;
    }

    @Override
    protected InferResult infer(int nrOfUnbound, InferContext context) {
        return context.shallow() ? unresolvable() : resolve(context.toDeep());
    }

    @Override
    public final InferResult resolve(InferContext context) {
        Predicate predicate = predicate();
        InferResult predResult = predicate.resolve(context);
        if (predResult.hasStackOverflow()) {
            return predResult;
        }
        return resolve(context, predResult);
    }

    protected abstract InferResult resolve(InferContext context, InferResult predResult);

}

File: web-ts/src/logic/Quantifier.ts
/**
 * Quantifier - base class for quantified predicates.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { Predicate } from './Predicate';
import { CompoundPredicate } from './CompoundPredicate';
import { InferResult } from '../InferResult';
import type { InferContext } from '../InferContext';

/**
 * Quantifier - base for existential and universal quantification.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier
 */
export abstract class Quantifier extends CompoundPredicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...args);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#doGetBinding(Object, int)
  protected override doGetBinding(_varVal: unknown, i: number): boolean {
    return i > 0;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#doSetBinding(Object, int)
  protected override doSetBinding(varVal: unknown, i: number): boolean {
    return i > 0 || varVal instanceof Variable;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#localVars()
  override localVars(): List<Variable> {
    return this.get(0) as List<Variable>;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#predicate()
  predicateBody(): Predicate {
    return this.predicateAt(1);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#countNrOfUnbound()
  protected override countNrOfUnbound(): number {
    const binding = this.getBinding();
    if (binding === null) return 0;
    const localVarsList = this.localVars();
    let count = 0;
    binding.forEach((val, key) => {
      if (val instanceof Type && !localVarsList.contains(key)) {
        count++;
      }
    });
    return count;
  }

  /**
   * Override inferInternal matching Java's Quantifier.infer(int, InferContext).
   * @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#infer(int, InferContext)
   */
  protected override inferInternal(_nrOfUnbound: number, context: InferContext): InferResult {
    return context.shallow() ? this.unresolvable() : this.resolve(context.toDeep());
  }

  /**
   * Override resolve matching Java's Quantifier.resolve(InferContext).
   * @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#resolve(InferContext)
   */
  override resolve(context: InferContext): InferResult {
    const predicate = this.predicateBody();
    const predResult = predicate.resolve(context);
    if (predResult.hasStackOverflow()) {
      return predResult;
    }
    return this.resolveWithResult(context, predResult);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Quantifier#resolve(InferContext, InferResult)
  protected abstract resolveWithResult(context: InferContext, predResult: InferResult): InferResult;
}
